keps/sig-storage/3107-csi-nodeexpandsecret/README.md

```diff
@@ -0,0 +1,278 @@
+
+CSI volumes that require secrets for online expansion will have NodeExpandSecretRef
+field set. If not set NodeExpandVolume CSI RPC call will be made without secret.
+Existing validation of PersistentVolume object will be relaxed to allow setting of
+NodeExpandSecretRef for the first time so as CSI volume expansion can be supported
+for existing PVs.
+
+CSI Spec 1.5 has added below field to facilitate and enable COs to make use of the
+same as part of the NodeExpandSecret
+
+```
+message NodeExpandVolumeRequest {
+  ...
+  // Secrets required by plugin to complete node expand volume request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 6
+    [(csi_secret) = true, (alpha_field) = true];
+}
+```
+The same field will be used by Kubernetes to fill secretRef in the
+NodeExpandVolume request.
+
+### Test Plan
+- Unit tests around all the added logic in kubelet.
+- Unit tests around all the added logic in kube-apiserver.
+- E2E tests around nodeExpandVolume to make sure the field value has passed
+  and can be used.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Implemented the feature.
+- Wrote all the unit and E2E tests.
+
+#### Beta
+
+- Deployed the feature in production and went through at least minor k8s
+  version.
+
+#### GA
+
+#### Deprecation
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+- **How can this feature be enabled / disabled in a live cluster?**
+
+  - Feature gate name: CSINodeExpandSecret
+  - Components depending on the feature gate: kubelet, kube-apiserver
+  - Will enabling / disabling the feature require downtime of the control
+    plane? no.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? yes.
+
+- **Does enabling the feature change any default behavior?** no.
+
+- **Can the feature be disabled once it has been enabled (i.e. can we roll
+  back the enablement)?** yes, if rollback of feature gate happened with the
+  field `NodeExpandRequest` set, it will exist, but be ignored.
+
+- **What happens if we reenable the feature if it was previously rolled
+  back?** nothing, as long as the new fields in `NodeExpandRequest` is not used.
+
+- **Are there any tests for feature enablement/disablement?** yes, unit tests
+  will cover this.
+
+### Rollout, Upgrade and Rollback Planning
+
+TBD
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+TBD
+
+###### What specific metrics should inform a rollback?
+
+TBD
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+TBD
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+TBD
+
+### Monitoring Requirements
```

Comment:
While not required for alpha, I encourage you to think about what metrics you'll need to build for beta.

Reply:
Sure , Thanks @deads2k :+1: